File: Difficulty: Medium/Next Greater Element in Circular Array/next-greater-element-in-circular-array.java
class Solution {
    public ArrayList<Integer> nextGreater(int[] arr) {
        // code here
        int n = arr.length;
        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            res[i] = -1;
        }

        Stack<Integer> st = new Stack<>();
        
        for (int i = 2 * n - 1; i >= 0; i--) {
            int curr = arr[i % n];

            // Pop elements smaller 
            // than or equal to current
            while (!st.isEmpty() && st.peek() <= curr) {
                st.pop();
            }

            // Assign next greater 
            // if within first pass
            if (i < n && !st.isEmpty()) {
                res[i] = st.peek();
            }

            st.push(curr);
        }

        ArrayList<Integer> result = new ArrayList<>();
        for (int val : res) {
            result.add(val);
        }

        return result;
    }
}

File: Difficulty: Medium/Next Greater Element in Circular Array/next-greater-element-in-circular-array.cpp
class Solution {
  public:
    vector<int> nextGreater(vector<int> &arr) {
        int n = arr.size();
        vector<int> result(n, -1); // Initialize result with -1 (default when no greater element found)
        stack<int> st; // Stack to keep track of potential next greater elements
        
        // Process the array in circular manner - traverse from right to left twice
        for(int i = 2 * n - 1; i >= 0; i--) {
            int idx = i % n; // Convert circular index to actual array index
            
            // Remove all elements from stack that are smaller than or equal to current element
            // We only want to keep elements that are strictly greater than current
            while(!st.empty() && st.top() <= arr[idx]) {
                st.pop();
            }
            
            // If stack is not empty, top element is the next greater element
            if(!st.empty()) {
                result[idx] = st.top();
            }
            
            // Push current element to stack for future comparisons
            st.push(arr[idx]);
        }
        return result;
    }
};